File: java/ISmsClient.java
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.StringJoiner;

public class ISmsClient {
    private final String apiKey;
    private final String baseUrl;
    private final HttpClient httpClient;

    public ISmsClient(String apiKey) {
        this(apiKey, "https://www.i-sms.app");
    }

    public ISmsClient(String apiKey, String baseUrl) {
        if (apiKey == null || apiKey.isBlank()) {
            throw new IllegalArgumentException("apiKey is required");
        }
        this.apiKey = apiKey;
        this.baseUrl = trimRightSlash(baseUrl);
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(20))
                .build();
    }

    public String searchProjects(String keyword) throws IOException, InterruptedException {
        return get("/api/v2/projects", Map.of("keyword", keyword));
    }

    public String getNumber(GetNumberRequest request) throws IOException, InterruptedException {
        Map<String, String> params = new LinkedHashMap<>();
        params.put("project_id", request.projectId);
        params.put("project_name", request.projectName);
        params.put("project_token", request.projectToken);
        params.put("quantity", String.valueOf(request.quantity <= 0 ? 1 : request.quantity));
        params.put("phone", request.phone);
        params.put("province", request.province);
        params.put("carrier", request.carrier);
        params.put("ascription", request.ascription);
        return get("/api/v2/get_number", params);
    }

    public String getSms(String orderId) throws IOException, InterruptedException {
        return getSms(orderId, null, null);
    }

    public String getSms(String orderId, String phoneNumber, String projectId) throws IOException, InterruptedException {
        Map<String, String> params = new LinkedHashMap<>();
        params.put("order_id", orderId);
        params.put("phone_number", phoneNumber);
        params.put("project_id", projectId);
        return get("/api/v1/get_sms", params);
    }

    public String releaseNumber(String orderId) throws IOException, InterruptedException {
        return releaseNumber(orderId, null, null);
    }

    public String releaseNumber(String orderId, String phoneNumber, String projectId) throws IOException, InterruptedException {
        Map<String, String> params = new LinkedHashMap<>();
        params.put("order_id", orderId);
        params.put("phone_number", phoneNumber);
        params.put("project_id", projectId);
        return get("/api/v1/release_number", params);
    }

    public String getUserInfo() throws IOException, InterruptedException {
        return get("/api/v1/user/info", Map.of());
    }

    private String get(String path, Map<String, String> params) throws IOException, InterruptedException {
        String query = buildQuery(params);
        String url = baseUrl + path + (query.isEmpty() ? "" : "?" + query);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(20))
                .header("Accept", "application/json")
                .header("X-API-KEY", apiKey)
                .GET()
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        return response.body();
    }

    private static String buildQuery(Map<String, String> params) {
        StringJoiner joiner = new StringJoiner("&");
        for (Map.Entry<String, String> entry : params.entrySet()) {
            String value = entry.getValue();
            if (value != null && !value.isBlank()) {
                joiner.add(encode(entry.getKey()) + "=" + encode(value));
            }
        }
        return joiner.toString();
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private static String trimRightSlash(String value) {
        return value == null ? "" : value.replaceAll("/+$", "");
    }

    public static class GetNumberRequest {
        public String projectId;
        public String projectName;
        public String projectToken;
        public int quantity = 1;
        public String phone;
        public String province;
        public String carrier;
        public String ascription;
    }
}

File: csharp/ISmsClient.cs
using System.Text.Json;

namespace ISms;

public sealed class ISmsClient
{
    private readonly string _apiKey;
    private readonly string _baseUrl;
    private readonly HttpClient _httpClient;

    public ISmsClient(string apiKey, string baseUrl = "https://www.i-sms.app", HttpClient? httpClient = null)
    {
        if (string.IsNullOrWhiteSpace(apiKey))
        {
            throw new ArgumentException("apiKey is required", nameof(apiKey));
        }

        _apiKey = apiKey;
        _baseUrl = baseUrl.TrimEnd('/');
        _httpClient = httpClient ?? new HttpClient { Timeout = TimeSpan.FromSeconds(20) };
    }

    public Task<JsonElement> SearchProjectsAsync(string keyword, CancellationToken cancellationToken = default)
    {
        return GetAsync("/api/v2/projects", new Dictionary<string, string?>
        {
            ["keyword"] = keyword
        }, cancellationToken);
    }

    public Task<JsonElement> GetNumberAsync(GetNumberRequest request, CancellationToken cancellationToken = default)
    {
        return GetAsync("/api/v2/get_number", new Dictionary<string, string?>
        {
            ["project_id"] = request.ProjectId,
            ["project_name"] = request.ProjectName,
            ["project_token"] = request.ProjectToken,
            ["quantity"] = (request.Quantity <= 0 ? 1 : request.Quantity).ToString(),
            ["phone"] = request.Phone,
            ["province"] = request.Province,
            ["carrier"] = request.Carrier,
            ["ascription"] = request.Ascription
        }, cancellationToken);
    }

    public Task<JsonElement> GetSmsAsync(string? orderId = null, string? phoneNumber = null, string? projectId = null, CancellationToken cancellationToken = default)
    {
        return GetAsync("/api/v1/get_sms", new Dictionary<string, string?>
        {
            ["order_id"] = orderId,
            ["phone_number"] = phoneNumber,
            ["project_id"] = projectId
        }, cancellationToken);
    }

    public Task<JsonElement> ReleaseNumberAsync(string? orderId = null, string? phoneNumber = null, string? projectId = null, CancellationToken cancellationToken = default)
    {
        return GetAsync("/api/v1/release_number", new Dictionary<string, string?>
        {
            ["order_id"] = orderId,
            ["phone_number"] = phoneNumber,
            ["project_id"] = projectId
        }, cancellationToken);
    }

    public Task<JsonElement> GetUserInfoAsync(CancellationToken cancellationToken = default)
    {
        return GetAsync("/api/v1/user/info", new Dictionary<string, string?>(), cancellationToken);
    }

    private async Task<JsonElement> GetAsync(string path, Dictionary<string, string?> parameters, CancellationToken cancellationToken)
    {
        var query = string.Join("&", parameters
            .Where(item => !string.IsNullOrWhiteSpace(item.Value))
            .Select(item => $"{Uri.EscapeDataString(item.Key)}={Uri.EscapeDataString(item.Value!)}"));

        var url = $"{_baseUrl}{path}{(query.Length == 0 ? string.Empty : "?" + query)}";
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Accept.ParseAdd("application/json");
        request.Headers.Add("X-API-KEY", _apiKey);

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        using var document = JsonDocument.Parse(body);
        return document.RootElement.Clone();
    }
}

public sealed class GetNumberRequest
{
    public string ProjectId { get; init; } = "";
    public string ProjectName { get; init; } = "";
    public string ProjectToken { get; init; } = "";
    public int Quantity { get; init; } = 1;
    public string? Phone { get; init; }
    public string? Province { get; init; }
    public string? Carrier { get; init; }
    public string? Ascription { get; init; }
}
